### PR TITLE
Aider/testing

### PR DESCRIPTION
**Pull request by kk60503**
I've added unit tests and some Read the Docs files, and have done a pretty large refactor of the repo. There are still some bugs I need to fix with the Read the Docs files, and I was not able to get the sql_agent fully working, but I think I got close. I was also not able to do extensive testing on any changes made. Most of the changes to pre-existing code such as extended exception handling and security was made by Aider.

I will keep working to get it fully functional, but I would like to get eyes on the changes and merge to `main` ASAP, since this is a pretty big overhaul.

### CODE REVIEW

**Comment by kk60503** [comment]
Also, the commits are kind of a mess from using Aider and also iteratively testing Read the Docs (sorry about that). I will see if I can get them squashed somewhat reasonably.

**Comment by QuriousMachine** [comment]
Hi Kyle
I've rarely done code review, can you remind me what are the things I need
to do? Thanks.
JunFan

**Comment by kk60503** [comment]
@CsTeSc In this case, I'm just looking for you guys to test and approve the changes that have been made to the modules/code you wrote.

File: tests/test_main.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        "chat_history": [],
+        "context": "test context",
+        "answer": "",
+    }
+
+
+@pytest.fixture
+def mock_rag_chain():
+    """Mock RAG chain fixture"""
+    with patch("langchain_app.base_model.rag_chain") as mock:
```
Comment (QuriousMachine) [inline tests/test_main.py]: Test fail at this line with follow error message:

name = 'langchain_app', import_ = <function _gcd_import at 0x7c18209d00e0>
>   ???
E   ModuleNotFoundError: No module named 'langchain_app'

<frozen importlib._bootstrap>:1324: ModuleNotFoundError
======== short test summary info =============
ERROR tests/test_main.py::test_call_model - ModuleNotFoundError: No module named 'langchain_app'

File: tests/test_sql_agent.py
Comment (QuriousMachine) [inline tests/test_sql_agent.py:1]: Multiple test failed with following error message:

FAILED tests/test_sql_agent.py::TestSQLAgent::test_load_docs - AssertionError: 0 != 2
FAILED tests/test_sql_agent.py::TestSQLAgent::test_load_docs_with_malicious_content - AssertionError: 0 != 1
FAILED tests/test_sql_agent.py::TestSQLAgent::test_load_urls - AssertionError: 0 != 1

File: tests/test_main.py (hunk 1 above)
Comment (kk60503) [reply tests/test_main.py]: Whoops. This one should be fixed now.

File: tests/test_sql_agent.py
Comment (kk60503) [reply tests/test_sql_agent.py:1]: Yeah, the SQL stuff is a lot more difficult to implement. We can probably just leave that unit test file out of the merge for now. I'll keep working on it if Grace doesn't get a chance to.

**Comment by kk60503** [comment]
Also, Aider has been writing Pytest functionality randomly into the unit tests, which I guess isn't a big deal but kinda weird to have both. They both should work, though.
Example `pytest` command:
`python -m pytest tests/test_main.py -v`

**Pull request closed without merging**